### PR TITLE
Add SnapshotLifecycleService and related CRUD APIs

### PR DESCRIPTION
**Pull request by dakrone**
This commit adds `SnapshotLifecycleService` as a new service under the ilm
plugin. This service handles snapshot lifecycle policies by scheduling based on
the policies defined schedule.

This also includes the get, put, and delete APIs for these policies

Relates to #38461

### CODE REVIEW

**Comment by dakrone** [comment]
I opened this as a `>non-issue` because once it is approved I intend to create a feature branch for this, rather than actually merging it to master. This is only a start of the framework for snapshot lifecycle management (and isn't feature-complete), and I'm opening it so we can start by having smaller reviewable chunks of the feature.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java
Comment (gwbrown) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java]: TODO: is this immutable?

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java
Comment (gwbrown) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java]: If we add the job using `computeIfAbsent` on a `ConcurrentHashMap`, we should probably remove it using `computeIfPresent`, e.g.:
```
scheduledTasks.computeIfPresent(snapshotLifecycleId, (id, job) -> {
  scheduler.remove(snapshotLifecycleId);
  return null;
});
```
That will make sure we have the same atomicity guarantees coming and going.

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/action/GetSnapshotLifecycleAction.java
Comment (gwbrown) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/action/GetSnapshotLifecycleAction.java]: Why not put this in a `toXContent` method on `SnapshotLifecyclePolicyItem`?

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java
Comment (gwbrown) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java]: It might be good for this to have a test similar to `IndexLifecycleServiceTests#prepareServices` to make sure it attaches the listener, but I'd also be fine leaving that as a TODO until there's something real to test.

**Review by AthenaEryma — CHANGES_REQUESTED** [review]
Left some comments - this looks like a good skeleton, though.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java
Comment (dakrone) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java]: Yep, I'll change this.

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecyclePolicy.java
Comment (dakrone) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecyclePolicy.java]: I'll use Object.equals for now, I don't want to add too much validation because I expect it to change as we work on it.

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java
Comment (dakrone) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java]: I don't think there's a preference. We use `newConcurrentMap` multiple places all over the code, so I stuck with that.

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java
Comment (dakrone) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java]: The `remove` for the scheduler is idempotent so I erred on the side of making sure to absolute remove the job even if the task map was out of sync.

I fully expect the implementation for this in particular (the map of tasks) to change, since right now it doesn't handle updates :)

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java
Comment (dakrone) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java]: Oh totally, I am planning on testing all of this, right now though, I think the implementation may change, so I'll leave it as a future TODO

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/action/GetSnapshotLifecycleAction.java
Comment (dakrone) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/action/GetSnapshotLifecycleAction.java]: Sure I can do that.

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleMetadata.java
Comment (dakrone) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleMetadata.java]: Removed this

**Comment by dakrone** [comment]
@gwbrown thanks for taking a look. I pushed commits addressing your feedback, can you take another look please?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java
Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java]: Switching this to `Collections.unmodifiableSet(schedules.keySet());` doesn't quite go far enough - the returned collection can't be modified, which solves that problem, but it's just a view into the underlying collection, not a copy, so iterating over it while removing items still hits the same undefined case.  Whether we want to make a copy in this method or at the site of the consumer I'm not sure - I'd err on the side of caution and do it here, but I could see the other way too.

**Comment by gwbrown** [comment]
Thanks @dakrone, most of the changes look good but I left one more comment on the `keySet` thing.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java
Comment (dakrone) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java]: > iterating over it while removing items still hits the same undefined case

Not sure I follow here, can you clarify what you mean? From the `unmodifiableSet` javadocs:

> Query operations on the returned set "read through" to the specified set, and attempts to modify the returned set, whether direct or via its iterator, result in an `UnsupportedOperationException`.

Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java]: Right, I'm looking at [`cancelSnapshotJobs()`](https://github.com/elastic/elasticsearch/pull/39795/files#diff-65a7e37d0dc6108c8613c07d4315ec9eR102), which calls `scheduledJobIds()`, then iterates over the returned set (which is a view into `schedules`), calling `remove()` on each one, which removes it from `schedules`.

So we're iterating over an unmodifiable view into `schedules` while removing things from `schedules`, which is exactly this case from the `keySet()` documentation: `If the map is modified while an iteration over the set is in progress (except through the iterator's own remove operation), the results of the iteration are undefined.`.

Fundamentally the problem here is that `Collections.unmodifiableX` is (unintuitively) just a view, not a copy - the returned instance can't be used to modify it, but if something else is holding on to the original object, it can be mutated and the "unmodifiable" instance will change.

E.g.:
```
        Map<String, String> test = new ConcurrentHashMap<>();
        test.put("key1", "val1");
        test.put("key2", "val2");
        Set<String> immutableTest = Collections.unmodifiableSet(test.keySet());
        test.remove("key1");

        System.out.print("test: ");
        System.out.println(test);

        System.out.print("immutableTest: ");
        System.out.println(immutableTest);
```
prints:
```
test: {key2=val2}
immutableTest: [key2]
```

Comment (dakrone) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java]: Ahh okay, I thought you were referring to something like this:

```java
        Iterator<String> i = scheduler.scheduledJobIds().iterator();
        while (i.hasNext()) {
            i.remove(); // <-- this "remove"
            i.next();
        }
```

Which is prevented, rather than the `remove` on the `SchedulerEngine`'s internal map.

I'll just copy the set.

**Comment by dakrone** [comment]
@gwbrown okay, I pushed a commit copying the set.

**Review by AthenaEryma — APPROVED** [review]
LGTM (assuming CI is still happy). Thanks!

**Comment by dakrone** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by dakrone** [comment]
Pushed branch "snapshot-lifecycle-management"

**Pull request closed without merging**
